File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/paths/BlueClose12.java
package org.firstinspires.ftc.teamcode.paths;

import com.pedropathing.follower.Follower;
import com.pedropathing.geometry.BezierCurve;
import com.pedropathing.geometry.BezierLine;
import com.pedropathing.geometry.Pose;
import com.pedropathing.paths.PathChain;

public class BlueClose12 {
    public PathChain ScorePreloads;
    public PathChain IntakeSpike1;
    public PathChain ScoreSpike1;
    public PathChain AlignSpike2;
    public PathChain IntakeSpike2;
    public PathChain ScoreSpike2;
    public PathChain PrepAlignSpike3;
    public PathChain IntakeSpike3;
    public PathChain ScoreSpike3Part1;
    public PathChain ScoreSpike3Part2;
    public PathChain Leave;

    public BlueClose12(Follower follower) {
        ScorePreloads = follower.pathBuilder().addPath(
                        new BezierCurve(
                                new Pose(32.000, 136.500),
                                new Pose(32.000, 107.000),
                                new Pose(42.000, 104.000),
                                new Pose(55.000, 88.000)
                        )
                ).setTangentHeadingInterpolation()
                .setReversed()
                .build();

        IntakeSpike1 = follower.pathBuilder().addPath(
                        new BezierCurve(
                                new Pose(55.000, 88.000),
                                new Pose(55.000, 84.000),
                                new Pose(18.000, 84.000)
                        )
                ).setConstantHeadingInterpolation(Math.toRadians(180))

                .build();

        ScoreSpike1 = follower.pathBuilder().addPath(
                        new BezierCurve(
                                new Pose(18.000, 84.000),
                                new Pose(55.000, 84.000),
                                new Pose(55.000, 88.000)
                        )
                ).setLinearHeadingInterpolation(Math.toRadians(180), Math.toRadians(129))

                .build();

        AlignSpike2 = follower.pathBuilder().addPath(
                        new BezierLine(
                                new Pose(55.000, 88.000),

                                new Pose(55.000, 62.0)
                        )
                ).setLinearHeadingInterpolation(Math.toRadians(129), Math.toRadians(180))

                .build();

        IntakeSpike2 = follower.pathBuilder().addPath(
                        new BezierLine(
                                new Pose(55.000, 62.0),

                                new Pose(18.000, 62.0)
                        )
                ).setTangentHeadingInterpolation()

                .build();

        ScoreSpike2 = follower.pathBuilder().addPath(
                        new BezierLine(
                                new Pose(18.000, 62.0),

                                new Pose(55.000, 88.000)
                        )
                ).setLinearHeadingInterpolation(Math.toRadians(180), Math.toRadians(129))
                .build();

        PrepAlignSpike3 = follower.pathBuilder().addPath(
                        new BezierLine(
                                new Pose(55.000, 88.000),

                                new Pose(55.000, 82.000)
                        )
                ).setLinearHeadingInterpolation(Math.toRadians(129), Math.toRadians(270))
                .build();

        IntakeSpike3 = follower.pathBuilder().addPath(
                        new BezierCurve(
                                new Pose(55.000, 82.000),
                                new Pose(67.000, 35.000),
                                new Pose(44.000, 35.000),
                                new Pose(18.000, 36.000)
                        )
                ).setTangentHeadingInterpolation()

                .build();

        ScoreSpike3Part1 = follower.pathBuilder().addPath(
                        new BezierLine(
                                new Pose(18.000, 36.000),

                                new Pose(50.000, 81.000)
                        )
                ).setConstantHeadingInterpolation(Math.toRadians(235))

                .build();

        ScoreSpike3Part2 = follower.pathBuilder().addPath(
                        new BezierLine(
                                new Pose(50.000, 81.000),

                                new Pose(55.000, 88.000)
                        )
                ).setLinearHeadingInterpolation(Math.toRadians(235), Math.toRadians(129))

                .build();

        Leave = follower.pathBuilder().addPath(
                        new BezierLine(
                                new Pose(55.000, 88.000),

                                new Pose(45.000, 78.000)
                        )
                ).setLinearHeadingInterpolation(Math.toRadians(129), Math.toRadians(0))

                .build();
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/auto/BlueClose12.kt
package org.firstinspires.ftc.teamcode.opmode.auto

import com.arcrobotics.ftclib.command.InstantCommand
import com.arcrobotics.ftclib.command.ParallelCommandGroup
import com.arcrobotics.ftclib.command.SequentialCommandGroup
import com.arcrobotics.ftclib.command.WaitCommand
import com.pedropathing.geometry.Pose
import com.pedropathing.paths.PathChain
import com.qualcomm.robotcore.eventloop.opmode.Autonomous
import org.firstinspires.ftc.teamcode.command.auto.PedroCommand
import org.firstinspires.ftc.teamcode.command.intake.IntakeWithGateClosed
import org.firstinspires.ftc.teamcode.command.launcher.FeedLauncherArtifacts
import org.firstinspires.ftc.teamcode.command.launcher.ManualHood
import org.firstinspires.ftc.teamcode.command.launcher.ManuallyLaunch
import org.firstinspires.ftc.teamcode.hardware.Robot
import org.firstinspires.ftc.teamcode.opmode.template.AutoTemplate
import org.firstinspires.ftc.teamcode.paths.BlueClose12Paths

@Autonomous(name = "Blue Close 12")
class BlueClose12 : AutoTemplate(Pose(26.0, 130.0, Math.toRadians(143.0))) {
    val command by lazy {
        fun construct(pathChain: PathChain): PedroCommand {
            return PedroCommand(pathChain, follower)
        }

        SequentialCommandGroup(
            // score preloads
            ParallelCommandGroup(IntakeWithGateClosed(), ManuallyLaunch { 0.65 }, ManualHood { 0.194 }),
            construct(BlueClose12Paths(follower).ScorePreloads),
            FeedLauncherArtifacts(),
            WaitCommand(750),

            // intake spike 2
            IntakeWithGateClosed(),
            construct(BlueClose12Paths(follower).IntakeSpike2),
            WaitCommand(750),
            construct(BlueClose12Paths(follower).OpenGate),
            WaitCommand(500),
            construct(BlueClose12Paths(follower).PrepScoreSpike2),
            construct(BlueClose12Paths(follower).ScoreSpike2),
            FeedLauncherArtifacts(),
            WaitCommand(750),

            IntakeWithGateClosed(),
            construct(BlueClose12Paths(follower).IntakeSpike1),
            WaitCommand(750),
            construct(BlueClose12Paths(follower).PrepScoreSpike1),
            construct(BlueClose12Paths(follower).ScoreSpike1),
            FeedLauncherArtifacts(),
            WaitCommand(750),

            IntakeWithGateClosed(),
            construct(BlueClose12Paths(follower).IntakeSpike3),
            WaitCommand(750),
            construct(BlueClose12Paths(follower).PrepScoreSpike3),
            construct(BlueClose12Paths(follower).ScoreSpike3),
            FeedLauncherArtifacts(),
            WaitCommand(750),

            InstantCommand({ stop() })
        )
    }

    override fun initialize() {
        command
        follower.setStartingPose(start)
        Robot.Subsystems.launcher.targetHoodByScalar(0.375)
    }

    override fun start() {
        super.start()

        Robot.scheduler.schedule(
            SequentialCommandGroup(
                command
            ),
        )
    }

    override fun cycle() {
        super.cycle()
    }
}